File: backend/src/Barnabas.Application/Search/SearchListings/SearchPage.cs
namespace Barnabas.Application.Search.SearchListings;

/// <summary>
/// A page of results, with the term that produced them.
/// </summary>
/// <remarks>
/// The term travels back so the screen can state what was searched without holding it — which is
/// what <c>L2-052</c> asks for on a search that found nothing, when there is nothing else on the
/// page to say what happened.
/// </remarks>
public sealed record SearchPage(string Term, IReadOnlyList<SearchResultDto> Results, string? NextCursor);

File: frontend/projects/api/src/lib/models/search-page.ts
import { ListingKind } from './listing-kind';

/** One listing a search found. */
export interface SearchResult {
  readonly listingId: string;
  readonly kind: ListingKind;
  readonly title: string;
  readonly ownerDisplayName: string;
  readonly neighbourhood: string;
  readonly price: number | null;
}

/**
 * A page of results, with the term that produced them.
 *
 * The term comes back rather than being remembered by the screen, so a search that found nothing
 * can still say what was searched — which is the whole of what that screen has to work with.
 */
export interface SearchPage {
  readonly term: string;
  readonly results: readonly SearchResult[];
  readonly nextCursor: string | null;
}

/** What narrows a search. */
export interface SearchFilters {
  readonly kind?: ListingKind;
  readonly neighbourhood?: string;
  readonly maxPrice?: number;
}
